### ISSUE TITLE
Provide a way to dereference return types

### ISSUE THREAD

**Issue by atrauzzi** [Microsoft/TypeScript#17400]
_This may or may not be similar to what's happening in #6606. I'm not 100% sure, so for the sake of completeness, I'm documenting my use case here._

One thing that I'm looking to reduce in one of my personal libraries & projects is the verbosity of declaring react-redux dispatcher structures.  In using them, I need to be able to hint-in their influence to the `props` property of my components:

```
export interface Props {

    // ...
}

export interface Dispatchers {

    // ...
}

export class MyComponent extends React.Component<Props & Dispatchers, any> {

    // ...
}

export const My = connect(
    () => {}, 
    {
        // ...
    } as Dispatchers
);
```

This eventually gets quite cumbersome because you're having to create and maintain information about the dispatchers in multiple places. So instead, I figure it would be nice to be able to simply declare once which dispatchers you were using by importing them.  From then on, it would be nice if there was a way for TypeScript to just know any time I type `this.props.` that the return type is the dispatch functions.

So I created this interface:

```
interface Dispatchers<T> {
    [D in keyof T]: (...args: any[]) => any;
}
```

This then results in my example above becoming:

```
import { dispatcherOne } from "./Dispatchers/DispatcherOne.ts";
import { dispatcherTwo } from "./Dispatchers/DispatcherTwo.ts";
import { dispatcherThree } from "./Dispatchers/DispatcherThree.ts";

export interface Props {

    // ...
}

export const dispatchers = {
    dispatcherOne,
    dispatcherTwo,
    dispatcherThree,
};

export class MyComponent extends React.Component<Props & Dispatchers<typeof dispatchers>, any> {

    // ...
}

export const My = connect(
    () => {}, 
    dispatchers
);
```

Although still a bit wordy, this actually helps me cut down on a massive amount of boilerplate code I was originally maintaining in my projects.

There's only one catch...

At present, I have no way of dereferencing my dispatcher functions which are passed to `connect` as **function factories**.  What I actually want hinted on `this.props.` is not the type that I imported, but the type returned by the type I imported.

In theory, what I really want would be something like this:

```
interface Dispatchers<T> {
    [D in keyof T]: (...args: any[]) => typeof T[D]();
}
```

See the small little change I dreamed up there?  I don't know if that passes any kind of smell test for TypeScript conventionally, but what I'm trying to say is that the value is not `typeof T[D]`.  I want the value of `[D in keyof T]` to be what's returned by `typeof T[D]`!

Anyway, thanks for hearing this one out.  If it is in fact a duplicate, feel free to close and mark it as such, but I hope I've given a clear enough example of a legitimate need for this kind of functionality.

cc. https://github.com/atrauzzi/protoculture-react/issues/5

**Comment by DanielRosenwasser** [comment]
Can you make it a little clearer what a dispatcher looks like, and why the mapped type needs to change the shape of your dispatchers object?

I'm familiar with some common patterns in Redux, but it's not exactly clear why this is necessary.

**Comment by atrauzzi** [comment]
Dispatchers look like this:

```
export function myDispatcher (dispatch, ownProps) {

    return (actual, parameters, go, here) => {

        // This is actually what gets called when you do `this.props.*` in a component.
    }
}
```

As per [react-redux](https://github.com/reactjs/react-redux).  And these [docs here](http://redux.js.org/docs/basics/UsageWithReact.html).  The `mapDispatchToProps` parameter in `connect` takes a dictionary of function factories.  It's the the functions returned by those function factories that end up being assigned to `this.props` back in React.

**Comment by mhegazy** [comment]
should be covered in #6606.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
